File: fast_barcode_scanner/android/src/main/kotlin/com/jhoogstraat/fast_barcode_scanner/types/ScannerConfiguration.kt
package com.jhoogstraat.fast_barcode_scanner.types

import android.util.Size
import com.google.mlkit.vision.barcode.Barcode

data class ScannerConfiguration(val formats: IntArray, val mode: DetectionMode, val resolution: Resolution, val framerate: Framerate, val position: CameraPosition)

enum class Framerate {
    fps30, fps60, fps120, fps240;

    fun intValue() : Int = when(this) {
        fps30 -> 30
        fps60 -> 60
        fps120 -> 120
        fps240 -> 240
    }

    fun duration() : Long = 1 / intValue().toLong()
}

enum class Resolution {
    sd480, hd720, hd1080, hd4k;

    private fun width() : Int = when(this) {
        sd480 -> 640
        hd720 -> 1280
        hd1080 -> 1920
        hd4k -> 3840
    }

    private fun height() : Int = when(this) {
        sd480 -> 480
        hd720 -> 720
        hd1080 -> 1080
        hd4k -> 2160
    }

    fun landscape() : Size = Size(width(), height())
    fun portrait() : Size = Size(height(), width())
}

enum class DetectionMode {
    pauseDetection, pauseVideo, continuous;
}

enum class CameraPosition {
    front, back;
}

val barcodeFormatMap = hashMapOf(
    "aztec" to Barcode.FORMAT_AZTEC,
    "code128" to Barcode.FORMAT_CODE_128,
    "code39" to Barcode.FORMAT_CODE_39,
    "code93" to Barcode.FORMAT_CODE_93,
    "codabar" to Barcode.FORMAT_CODABAR,
    "dataMatrix" to Barcode.FORMAT_DATA_MATRIX,
    "ean13" to Barcode.FORMAT_EAN_13,
    "ean8" to Barcode.FORMAT_EAN_8,
    "itf" to Barcode.FORMAT_ITF,
    "pdf417" to Barcode.FORMAT_PDF417,
    "qr" to Barcode.FORMAT_QR_CODE,
    "upcA" to Barcode.FORMAT_UPC_A,
    "upcE" to Barcode.FORMAT_UPC_E
)

val barcodeStringMap = barcodeFormatMap.entries.associateBy({ it.value }) { it.key }

File: fast_barcode_scanner/ios/Classes/types/ScannerConfiguration.swift
import AVFoundation
import Vision

struct ScannerConfiguration {

    init(position: AVCaptureDevice.Position,
         framerate: Framerate,
         resolution: Resolution,
         mode: DetectionMode,
         codes: [String]) {
        self.position = position
        self.framerate = framerate
        self.resolution = resolution
        self.detectionMode = mode
        self.codes = codes
    }

    init?(_ args: Any?) {
        guard
            let dict = args as? [String: Any],
            let position = cameraPositions[dict["pos"] as? String ?? ""],
            let resolution = Resolution(rawValue: dict["res"] as? String ?? ""),
            let framerate = Framerate(rawValue: dict["fps"] as? String ?? ""),
            let detectionMode = DetectionMode(rawValue: dict["mode"] as? String ?? ""),
            let codes = dict["types"] as? [String]
            else {
                return nil
        }

        self.init(position: position,
                  framerate: framerate,
                  resolution: resolution,
                  mode: detectionMode,
                  codes: codes
        )
    }

    let position: AVCaptureDevice.Position
    let framerate: Framerate
    let resolution: Resolution
    let detectionMode: DetectionMode
    let codes: [String]

    func copy(with args: Any?) -> ScannerConfiguration? {
        guard let dict = args as? [String: Any] else { return nil }

        return ScannerConfiguration.init(
            position: cameraPositions[dict["pos"] as? String ?? ""] ?? position,
            framerate: Framerate(rawValue: dict["fps"] as? String ?? "") ?? framerate,
            resolution: Resolution(rawValue: dict["res"] as? String ?? "") ?? resolution,
            mode: DetectionMode(rawValue: dict["mode"] as? String ?? "") ?? detectionMode,
            codes: dict["types"] as? [String] ?? codes
        )
    }
}

// Flutter -> AVFoundation
let avMetadataObjectTypes: [String: AVMetadataObject.ObjectType] =
[
    "aztec": .aztec,
    "code128": .code128,
    "code39": .code39,
    "code39mod43": .code39Mod43,
    "code93": .code93,
    "dataMatrix": .dataMatrix,
    "ean13": .ean13,
    "ean8": .ean8,
    "itf": .itf14,
    "pdf417": .pdf417,
    "qr": .qr,
    "upcE": .upce,
    "interleaved": .interleaved2of5
]

// Flutter -> Vision
@available(iOS 11, *)
let vnBarcodeSymbols: [String: VNBarcodeSymbology] =
[
    "aztec": .aztec,
    "code128": .code128,
    "code39": .code39, // Which one?
    "code93": .code93, // Which one?
    "dataMatrix": .dataMatrix,
    "ean13": .ean13,
    "ean8": .ean8,
    "itf": .itf14,
    "pdf417": .pdf417,
    "qr": .qr,
    "upcE": .upce,
    "interleaved": .i2of5 // Which one?
]

// AVFoundation -> Flutter
let flutterMetadataObjectTypes = Dictionary(uniqueKeysWithValues: avMetadataObjectTypes.map { ($1, $0) })

// Vision -> Flutter
@available(iOS 11, *)
let flutterVNSymbols = Dictionary(uniqueKeysWithValues: vnBarcodeSymbols.map { ($1, $0) })

let cameraPositions: [String: AVCaptureDevice.Position] =
[
    "front": .front,
    "back": .back
]

enum Resolution: String {
    case sd480, hd720, hd1080, hd4k

    var width: Int32 {
        switch self {
        case .sd480: return 720
        case .hd720: return 1280
        case .hd1080: return 1920
        case .hd4k: return 3840
        }
    }

    var height: Int32 {
        switch self {
        case .sd480: return 480
        case .hd720: return 720
        case .hd1080: return 1080
        case .hd4k: return 2160
        }
    }
}

enum Framerate: String {
    case fps30, fps60, fps120, fps240

    var doubleValue: Double {
        switch self {
        case .fps30: return 30
        case .fps60: return 60
        case .fps120: return 120
        case .fps240: return 240
        }
    }
}

enum DetectionMode: String {
    case pauseDetection, pauseVideo, continuous
}
